### PR TITLE
many: drop SnapFile and snapYaml usage in early install

### PR DESCRIPTION
**Pull request by pedronis**
This branch does accomplish:

* gets rid of all usages of snapYaml in the early phase of install (where we were using SnapFile) in favor of snap.Info
* in the process moves legacy bits of snap.yaml syntax under Info.Legacy which just receives the parsed stuff without further processing (see snap/legacy.go, gadget yaml defs are moved under snap/legacygadget, processing is still in snappy/)
* drops some usages of NewSnapFile, for the rest gets rid of SnapFile and use openSnapBlob instead taking a path and a snap.SideInfo and returning a Info and snap.File
* use Info.MountFile (instead of BlobPath, make squashfs not need to know about snappy/dirs) and rename Info.BaseDir to Info.MountDir
* other cleanups

given that we still have snap.Snap for now there is some ugliness at the points where the code isn't moved yet to use Infos, especially when the two approaches meet, but still should set up things enough to be able to switch to use Revision for building disk paths. we can proceed with what to do with snap.Snap after.

There are still some vestigial uses of snapYaml which can also be dealt with later, and in security.go which should just die(tm).

### CODE REVIEW

File: snap/file.go [hunk 1]
```diff
@@ -27,7 +27,8 @@ import (
 
 // File is the interface to interact with the low-level snap files
 type File interface {
-	Install(targetDir string) error
+	// Install copies the snap file to targetPath (and possibly unpacks it to mountDir)
```
Comment (niemeyer) [inline snap/file.go]: s/unpacks/mounts/

File: snappy/snap_file.go [hunk 2]
```diff
@@ -20,74 +20,29 @@
...
-}
+// openSnapBlob opens a snap blob returning both a snap.Info completed
+// with sideInfo (if not nil) and a corresponding snap.File.
+func openSnapBlob(snapFile string, unsignedOk bool, sideInfo *snap.SideInfo) (*snap.Info, snap.File, error) {
+	// TODO: what precautions to take if unsignedOk == false ?
 
-// NewSnapFile loads a snap from the given snapFile
-func NewSnapFile(snapFile string, unsignedOk bool) (*SnapFile, error) {
-	d, err := snap.Open(snapFile)
+	blobf, err := snap.Open(snapFile)
```
Comment (niemeyer) [inline snappy/snap_file.go]: I suggest renaming blobf to snapf everywhere. Commenting here because this seems like the most unclear case, since we're getting a snapFile as a parameter and returning a snap.File. For clarity, I'd rename:
* openSnapBlob => openSnapFile
* snapFile => snapPath
* blobf => snapf (here and everywhere)

**Comment by niemeyer** [comment]
LGTM! Just one suggestion.

File: snap/file.go (hunk 1 above)
Comment (pedronis) [inline snap/file.go]: I think the corner case is really unpacks (maybe I should just leave it out), it's used by tests, maybe @mvo5 know more

File: snap/info.go [hunk 3]
```diff
@@ -89,11 +93,16 @@ func (s *Info) Description() string {
...
 
-// BaseDir returns the base directory of the snap.
-func (s *Info) BaseDir() string {
+// MountDir returns the base directory of the snap where it gets mounted.
+func (s *Info) MountDir() string {
 	return filepath.Join(dirs.SnapSnapsDir, s.Name(), s.Version)
 }
 
+// MountFile returns the path where the snap file that is mounted is installed.
+func (s *Info) MountFile() string {
```
Comment (mvo5) [inline snap/info.go]: :heart:

File: snap/legacy.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package snap
+
+import (
+	"github.com/ubuntu-core/snappy/snap/legacygadget"
+)
+
+// LegacyYaml collects the legacy fields in snap.yaml that are up to be reworked.
+type LegacyYaml struct {
```
Comment (mvo5) [inline snap/legacy.go]: Nice, I like that!

File: snap/squashfs/squashfs.go [hunk 5]
```diff
@@ -72,10 +60,10 @@ func (s *Snap) MetaMember(metaMember string) ([]byte, error) {
 }
 
 // Install just copies the blob into place (unless it is used in the tests)
-func (s *Snap) Install(instDir string) error {
+func (s *Snap) Install(targetPath, instDir string) error {
```
Comment (mvo5) [inline snap/squashfs/squashfs.go]: Maybe `mountDir` instead of `instDir` for the parameter here to make it more clear what the difference is?

File: snappy/kernel_os.go [hunk 6]
```diff
@@ -67,9 +66,9 @@ func removeKernelAssets(s *Snap, inter interacter) error {
 // extractKernelAssets extracts kernel/initrd/dtb data from the given
 // Snap to a versionized bootloader directory so that the bootloader
 // can use it.
-func extractKernelAssets(s *SnapFile, inter progress.Meter, flags InstallFlags) error {
-	if s.m.Type != snap.TypeKernel {
-		return fmt.Errorf("can not extract kernel assets from snap type %q", s.Type())
+func extractKernelAssets(s *snap.Info, snapf snap.File, flags InstallFlags, inter progress.Meter) error {
```
Comment (mvo5) [inline snappy/kernel_os.go]: Thanks for making the argument order here consistent!

File: snappy/snap_file.go [hunk 7]
```diff
@@ -20,74 +20,29 @@
...
-	"path/filepath"
-
-	"github.com/ubuntu-core/snappy/dirs"
 	"github.com/ubuntu-core/snappy/snap"
 )
 
-// SnapFile is a local snap file that can get installed
-type SnapFile struct {
-	m   *snapYaml
-	deb snap.File
```
Comment (mvo5) [inline snappy/snap_file.go]: \o/ For killing `deb` here :P

**Comment by mvo5** [comment]
Looks good! Thanks for doing this work.
